### PR TITLE
fix(eslint-plugin): [require-await] handle async generators

### LINKED ISSUE #1691: [require-await] False positive on async generator that uses yield*
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/require-await": "error"
  }
}
```

```TS
async function* foo() {
  await Promise.resolve()
  yield 1
}
async function* bar() {
  yield* foo()
}
```

**Expected Result**
No error. `bar` has to be async because it delegates to `foo`, which is async.

**Actual Result**
Error: Async generator function has no 'await' expression.

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.22.0` |
| `@typescript-eslint/parser`        | `X.Y.Z` |
| `TypeScript`                       | `X.Y.Z` |
| `ESLint`                           | `X.Y.Z` |
| `node`                             | `X.Y.Z` |
| `npm`                              | `X.Y.Z` |

### PR DESCRIPTION
**Pull request by anikethsaha**
Fixes #1691 

- added `!node.generator` before checking starts as to ignore the checks for generators
- added tests

### CODE REVIEW

**Comment by anikethsaha** [comment]
> To ensure that code like the following aren't considered valid:
> 
> ```
> async function* foo() {} // no await or async yield
> async function* bar() {
>   return 1; // not a promise
> }
> async function* baz() {
>   yield 1; // not an async yield
> }
> ```

These will be valid as these contains generators (along with async). Having generator will be a valid test irrespective of the function body 

https://github.com/typescript-eslint/typescript-eslint/pull/1782/files#diff-0c97f9870cc2985da0e0acb232680f9bR122

Let me know if I am wrong !

**Comment by bradzacher** [comment]
Sorry, what I'm saying is that they should not be valid.

Adding `*` should not make the function bypass all checks this rule already does, it should just add an additional allowance in the form of checking for, and not reporting when the function contains a `yield*`.

**Comment by anikethsaha** [comment]
Actually in eslint repo, there is a [similar issue](https://github.com/eslint/eslint/issues/12459) in which it states that it should not check for yield as well cause async generators cant be yield that are coming generator function. 

IMO, I would suggest having a separate rule like `require-yield` if not present currently

Let me know, if you want me to change this, I will add a check for `yield *` for generator function but I think its not in the scope of this rule to check yield

**Comment by bradzacher** [comment]
Note that you should always take the discussions in the ESLint repo with a pinch of salt in the context of this project, because we run in a typechecked world here.
Eg unlike the base repo, we can (and do) go one step further and deduce if a function returns a promise, making `async () => Promise.resolve()` valid in this rule.

----

Looking at it from this point of view, which is the question the rule is attempting to ask in a codebase: Does the async function require an await expression?

- an async function that returns a promise ***does not*** need an await.
   - as mentioned, this is one check we've added to the rule
- an async generator with just `yield 1` ***does*** need an await.
   - otherwise, it's just a normal generator.
- an async generator with `yield* syncGenerator()` ***does*** need an await.
   - otherwise, it's just a normal generator.
- an async generator with `yield* asyncGenerator()` ***does not*** need an await, because the yielded generator is an async generator.

The last case is why they don't want to check them in the base rule - they cannot detect this case. OTOH, we can check it in here, because we have type information.

So rather than just ignoring all async generators, we should just ignore async generators that `yield* asyncGenerator()`.

**Comment by anikethsaha** [comment]
> we can (and do) go one step further and deduce if a function returns a promise,

I completely forgot the power of typescript for this :sweat_smile:


> So rather than just ignoring all async generators, we should just ignore async generators that `yield* asyncGenerator()`

also for returning promises as well right ?

**Comment by bradzacher** [comment]
> also for returning promises as well right ?

Yes, sorry to clarify I meant this:
"in the case of async generators, we should _additionally_ ignore if we detect `yield* asyncGenerator()`"

**Comment by anikethsaha** [comment]
ok cool, I will make the necessary changes

**Comment by anikethsaha** [comment]
> * an async generator with `yield* syncGenerator()` _**does**_ need an await.
>   
>   * otherwise, it's just a normal generator.
> * an async generator with `yield* asyncGenerator()` _**does not**_ need an await, because the yielded generator is an async generator.

@bradzacher can you give an example of these two cases

**Comment by bradzacher** [comment]
```ts
async function* asyncGenerator() {
  await Promise.resolve()
  yield 1
}
function* syncGenerator() {
  yield 1
}

// OKAY - async generator deferring to an async generator
async function* test1() {
  yield* asyncGenerator()
}
// TS error - non-async generator deferring to an async generator
function* test2() {
  yield* asyncGenerator()
}

// LINT ERROR - async generator has no await and is not deferring to an async generator
async function* test3() {
  yield asyncGenerator()
}
// OKAY - no async
function* test4() {
  yield asyncGenerator()
}

// LINT ERROR - async generator has no await and is deferring to a non-async generator
async function* test5() {
  yield* syncGenerator()
}
// OKAY - no async
function* test6() {
  yield* syncGenerator()
}

// LINT ERROR - async generator has no await and is not deferring to an async generator
async function* test7() {
  yield syncGenerator()
}
// OKAY - no async
function* test8() {
  yield syncGenerator()
}
```
[repl](https://www.typescriptlang.org/v2/en/play?#code/JYOwLgpgTgZghgYwgAgMoE8C2AjA9gGwGFcQBnMKAVwTFymQG8AoZV5AegCpOW3lPkAQWSYIYABa4AJsglwwyKGMpQys8SikR4lfAril0IBMmCQo8ushhW4IZLmwArCDQB0yQnHz4IM7OjqKKQQmHZgwAikDjC8bAISKDZQALRwAO5wZim4MMjk8qEQ4G5xrJzsZUpwUiT4gQZGCACS5pZQAFz5WHj4ANxMAL5MTKDm8EhChsat0O0APAAqADTIiwBKyqrIALzIduiriwByEAAeCnuUIFowoH4AfIxl7OzIxwDyiwCiXQDkIHOYD+pmit3uMkopFAAHN9rJKAAHXyyXDIYqkFQodIoJSIugKRLIBB0JQ0fakaEwkBwbDAfBmQLQKB0aKgfY+ZDIxAQUilPiAi4ACjcorgUBhpC6AG0ALrIAA+yGlJyBsoAlF0AAoszDAELzWYWWhQTaYvRLI6bMAqEAPB4DPhKG2qAD8QoAbt5KBBXV0Nlt7Eqdbg9SEADLAADWECW1ttD01yBDYdjRvaZt0YEta3jqntjrYEhZ6Xdvq6ByTKf1abaJszFpWucDBaGIzG0AmKEE0xabWwviWT2YfGlGBwBDcjRmdbosqFSZ7TXTJqHA2Go3AnZ5U2X-d8K7oQ-RF2KUmiS5nc1Xi2HZTHPUn07717nC66l5fFgHEEPUDXbZMK8lTAcg3zHAAIsgADqHzrAA0oI6wfAAqpByAAGJwcg4hgGAiJSq8MJmOIlDYG4JKYOweoICypC5GA7CLOgiIQKgtHAIiYApNBEDYNCkDsPqmK8uwACMACsQGvCMslMM+1jXDQwAkAIz4AOLFK+UALs8rAZFkCjViEbhKPR+AehAC68OgwAQPgMhiW2MBKREqndMYmmAsadC6SOyC2fZjmAfJvaKcYbkgAkvJgGJfk2XZDlqb2XnadZwwuRFKlRbIMUAEzxawgVJRSTSpT5OnqoBCmZcp7mQOQADMhUBYlMgaVpFXpUwtWRdF5AACwtcV7UpZ17TdaFTThXVOUNWAEnDW1Ahjd5E1VRlrnZf1YAAGxLUFK1leNJqTTVW31TFADsB0OR5CDletzkXXNMUABy3TIq1pRtTBAA)

**Comment by anikethsaha** [comment]
Cool, I am adding these example as tests 👍

**Comment by anikethsaha** [comment]
- This should not be invalid for this rule right cause its not an async function so it wont be invalid
```js
function* test2() {
  yield* asyncGenerator()
}
```

- This should be Lint free right ? 

```js
async function* test5() {
  yield* syncGenerator()
}
```

**Comment by bradzacher** [comment]
The former is ignored by the rule, yes, as it's not async.

The latter should be a lint error:
```
// LINT ERROR - async generator has no await and is deferring to a non-async generator
```

**Comment by anikethsaha** [comment]
is `syncGenerator` and `asyncGenerator` are special functions ?  cause in AST it's simply an `identifer`

**Comment by anikethsaha** [comment]
whats the difference between 

```js
async function* test1() {
  yield* asyncGenerator()
}
```
and 

```js
async function* test1() {
  yield* syncGenerator()
}
```

**Comment by anikethsaha** [comment]
Ok got it, here is the defination, 

> ```ts
> // OKAY - async generator has an await
> async function* asyncGenerator() {
>   await Promise.resolve()
>   yield 1
> }
> // OKAY - no async
> function* syncGenerator() {
>   yield 1
> }
> ```

So, we need to check and test the function's declaration as well right to achieve this right ?

**Comment by anikethsaha** [comment]
cc @bradzacher

**Comment by bradzacher** [comment]
Yup! Similar to what we currently do for return statements.
With return statements - we grab the argument, and check to see if it's a promise.

With a `yield*` statement, we can similarly grab an argument, and check to see if its type is `AsyncGenerator`.

[A good place to play with this is ts-ast-viewer](https://ts-ast-viewer.com/#code/JYOwLgpgTgZghgYwgAgMoE8C2AjA9gGwGFcQBnMKAVwTFymQG8AoZV5AegCpOW3lPkAQWSYIYABa4AJsglwwyKGMpQys8SikR4lfAril0IBMmCQo8ushhW4IZLmwArCDQB0yQnHz4IM7OjqKKQQmHZgwAikDjC8bAISKDZQALRwAO5wZim4MMjk8qEQ4G5xrJzsZUpwUiT4gQZGCACS5pZQAFz5WHj4ANxMAL5MTKDm8EhChsat0O0APAAqADTIiwBKyqrIALzIduiriwByEAAeCnuUIFowoH4AfIxl7OzIxwDyiwCiXQDkIHOYD+pmit3uMkopFAAHN9rJKAAHXyyXDIYqkFQodIoJSIugKRLIBB0JQ0fakaEwkBwbDAfBmQLQKB0aKgfY+ZDIxAQUilPiAi4ACjcorgUBhpC6AG0ALrIAA+yGlJyBsoAlF0AAoszDAELzWYWWhQTaYvRLI6bMAqEAPB4DPhKG2qAD8QoAbt5KBBXV0Nlt7Eqdbg9SEADLAADWECW1ttD01yBDYdjRvaZt0YEta3jqntjrYEhZ6Xdvq6ByTKf1abaJszFpWucDBaGIzG0AmKEE0xabWwviWT2YfGlGBwBDcjRmdbosqFSZ7TXTJqHA2Go3AnZ5U2X-d8K7oQ-RF2KUmiS5nc1Xi2HZTHPUn07717nC66l5fFgHEEPUDXbZMK8lTAcg3zHAAIsgADqHzrAA0oI6wfAAqpByAAGJwcg4hgGAiJSq8MJmOIlDYG4JKYOweoICypC5GA7CLOgiIQKgtHAIiYApNBEDYNCkDsPqmK8uwACMACsQGvCMskjK8yAfIhACayApBSTTIDCxSvvQ4gGPs9gZFkYBMM+1jXDQwAkAIz4AOI6cadALs8rDGWYya6jWbhKPR+AehAC68OgwAQPgMhiW2ClKYIqnqSAaLPkwMCWRENndMYDmAk5UAuSOyAhWFEWAeZKXGGlIACDYuB5WUhXhbZvZZbpC6Fqw9VSAITWOe0rVDEAA)

If you select one of the call expressions inside the `foo` function, then put this in the console, you can see how to uniquely tell what the type's name is.

image

`type.target.getSymbol().getName()`

**Comment by anikethsaha** [comment]
> [A good place to play with this is ts-ast-viewer](https://ts-ast-viewer.com/#code/JYOwLgpgTgZghgYwgAgMoE8C2AjA9gGwGFcQBnMKAVwTFymQG8AoZV5AegCpOW3lPkAQWSYIYABa4AJsglwwyKGMpQys8SikR4lfAril0IBMmCQo8ushhW4IZLmwArCDQB0yQnHz4IM7OjqKKQQmHZgwAikDjC8bAISKDZQALRwAO5wZim4MMjk8qEQ4G5xrJzsZUpwUiT4gQZGCACS5pZQAFz5WHj4ANxMAL5MTKDm8EhChsat0O0APAAqADTIiwBKyqrIALzIduiriwByEAAeCnuUIFowoH4AfIxl7OzIxwDyiwCiXQDkIHOYD+pmit3uMkopFAAHN9rJKAAHXyyXDIYqkFQodIoJSIugKRLIBB0JQ0fakaEwkBwbDAfBmQLQKB0aKgfY+ZDIxAQUilPiAi4ACjcorgUBhpC6AG0ALrIAA+yGlJyBsoAlF0AAoszDAELzWYWWhQTaYvRLI6bMAqEAPB4DPhKG2qAD8QoAbt5KBBXV0Nlt7Eqdbg9SEADLAADWECW1ttD01yBDYdjRvaZt0YEta3jqntjrYEhZ6Xdvq6ByTKf1abaJszFpWucDBaGIzG0AmKEE0xabWwviWT2YfGlGBwBDcjRmdbosqFSZ7TXTJqHA2Go3AnZ5U2X-d8K7oQ-RF2KUmiS5nc1Xi2HZTHPUn07717nC66l5fFgHEEPUDXbZMK8lTAcg3zHAAIsgADqHzrAA0oI6wfAAqpByAAGJwcg4hgGAiJSq8MJmOIlDYG4JKYOweoICypC5GA7CLOgiIQKgtHAIiYApNBEDYNCkDsPqmK8uwACMACsQGvCMskjK8yAfIhACayApBSTTIDCxSvvQ4gGPs9gZFkYBMM+1jXDQwAkAIz4AOI6cadALs8rDGWYya6jWbhKPR+AehAC68OgwAQPgMhiW2ClKYIqnqSAaLPkwMCWRENndMYDmAk5UAuSOyAhWFEWAeZKXGGlIACDYuB5WUhXhbZvZZbpC6Fqw9VSAITWOe0rVDEAA)
> 
> If you select one of the call expressions inside the `foo` function, then put this in the console, you can see how to uniquely tell what the type's name is.


I am not able to get this thing with this site. 

[is it possible if you can show it using `AST-Explorer ?`](https://astexplorer.net/#/gist/365ff62c115bea930b2438c9408e974e/aa4837a8c95a026c204b9dee21e3e8ccff7d67ef)

Also,

this what the latest change looks like 

[`repl.it`](https://repl.it/@anikethsaha/typescript-eslint-1782)

Now , 

```typescript
async function* asyncGenerator() {
  await Promise.resolve()
  yield 1
}


async function* test1() {
  yield* asyncGenerator()
}
```

We need to get the definition of `asyncGenerator` and need to just check whether its `async` or not right ? for this , we need to somehow get the definaitino of this function and check it. Might increase the complexity of this rule. _(I might be wrong, I think I am missing something)_

**Comment by anikethsaha** [comment]
@bradzacher you can check the diff here with reference to this PR's previous commit 
https://github.com/typescript-eslint/typescript-eslint/pull/1782/commits/6d8c439ed0b73c8ad5a54f31ea89f2958c1c5953

**Comment by bradzacher** [comment]
Here's how to interact with the call expression in ts-ast-viewer:
ezgif-3-9eef18b01bb0
Sorry about the size, the free converter wouldn't let me do full resolution.


unfortunately, astexplorer does not show type information - it only shows the AST.
OTOH, ts-ast-viewer shows you type information.

So when working with type info rules you kinda have to use both as you learn and get familiar with it all.

Essentially you will want do do something like the following:
```
'YieldExpression[delegate=true]'(node: TSESTree.YieldExpression): void {
  const tsNode = parserServices.esTreeNodeToTSNodeMap.get(node.argument);
  const type = checker.getTypeAtLocation(tsNode);
  const symbol = type.getSymbol();
  if (symbol.getName() === 'AsyncGenerator' && scopeInfo) {
    scopeInfo.hasAsyncDelegateYield = true;
  }
},
```

**Comment by anikethsaha** [comment]
```ts
'YieldExpression[delegate=true]'(node: TSESTree.YieldExpression): void {
  const tsNode = parserServices.esTreeNodeToTSNodeMap.get(node.argument);
  const type = checker.getTypeAtLocation(tsNode);
  const symbol = type.getSymbol(); // <--- undefined
  if (symbol.getName() === 'AsyncGenerator' && scopeInfo) {
    scopeInfo.hasAsyncDelegateYield = true;
  }
},
```

@bradzacher `type.getSymbol();` and even `type.symbol` both are `undefined`

**Comment by bradzacher** [comment]
> type.getSymbol(); and even type.symbol both are undefined

This is expected - some types won't have an associated symbol.
In this case you can just ignore the ones that don't - as they won't ever be `AsyncGenerator` types.

```ts
if (symbol?.getName() === 'AsyncGenerator' && scopeInfo) {
```

**Comment by anikethsaha** [comment]
@bradzacher any suggestion for type checks errors !?

File: packages/eslint-plugin/src/rules/require-await.ts [hunk 1]
```diff
@@ -92,10 +119,34 @@ export default util.createRule({
...
       }
-
       scopeInfo.hasAwait = true;
     }
 
+    /**
+     * mark `scopeInfo.isAsyncYield` to `true` if its a generator
+     * function and the delegate is `true`
+     */
+    function markAsHasDelegateGen(node: FunctionNode): void {
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/require-await.ts]:
```suggestion
    function markAsHasDelegateGen(node: TESTree.YieldExpression): void {
```

File: packages/eslint-plugin/src/rules/require-await.ts [hunk 2]
```diff
@@ -92,10 +119,34 @@ export default util.createRule({
...
+        return;
+      }
+
+      if (node?.argument?.type === AST_NODE_TYPES.Literal) {
+        // making this `true` as for literals we dont need to check the defination
+        // eg : async function* run() { yield* 1 }
+        scopeInfo.isAsyncYield = true;
+      }
+
+      const tsNode = parserServices.esTreeNodeToTSNodeMap.get(node?.argument);
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/require-await.ts]:
```suggestion
      if (!scopeInfo || !scopeInfo.isGen || !node.argument) {
        return;
      }

      if (node.argument.type === AST_NODE_TYPES.Literal) {
        // making this `true` as for literals we dont need to check the defination
        // eg : async function* run() { yield* 1 }
        scopeInfo.isAsyncYield = true;
      }

      const tsNode = parserServices.esTreeNodeToTSNodeMap.get(node.argument);
```

File: packages/eslint-plugin/src/rules/require-await.ts [hunk 3]
```diff
@@ -49,9 +51,28 @@ export default util.createRule({
...
+          !!~['Promise'].indexOf(
+            node.returnType?.typeAnnotation?.typeName?.name,
+          )
+        ) {
+          return true;
+        }
+      }
+      return false;
+    }
+
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/require-await.ts]: Don't need this any more

```suggestion
```

File: packages/eslint-plugin/src/rules/require-await.ts [hunk 4]
```diff
@@ -62,7 +83,13 @@ export default util.createRule({
         return;
       }
 
-      if (node.async && !scopeInfo.hasAwait && !isEmptyFunction(node)) {
+      if (
+        !isValidAsyncGenerator(node) &&
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/require-await.ts]: ditto
```suggestion
```

**Pull request merged**